src/PrestaShopBundle/Form/Admin/Sell/Address/CustomerAddressType.php

```diff
@@ -0,0 +1,484 @@
+                        ),
+                    ]),
+                ],
+            ])
+            ->add('phone', TextType::class, [
+                'required' => $this->isRequired('phone'),
+                'constraints' => [
+                    new NotBlank(
+                        [
+                            'groups' => 'phone',
+                            'message' => $this->translator->trans(
+                                'This field cannot be empty', [], 'Admin.Notifications.Error'
+                            ),
+                        ]
+                    ),
+                    new CleanHtml(),
+                    new TypedRegex([
+                        'type' => 'phone_number',
+                    ]),
+                    new Length(
+                        [
+                            'max' => AddressConstraint::MAX_PHONE_LENGTH,
+                            'maxMessage' => $this->translator->trans(
+                                'This field cannot be longer than %limit% characters',
+                                ['%limit%' => AddressConstraint::MAX_PHONE_LENGTH],
+                                'Admin.Notifications.Error'
+                            ),
+                        ]
+                    ),
+                ],
+            ])
+            ->add('other', TextareaType::class, [
+                'required' => $this->isRequired('other'),
+                'constraints' => [
+                    new NotBlank(
+                        [
+                            'groups' => 'other',
+                            'message' => $this->translator->trans(
+                                'This field cannot be empty', [], 'Admin.Notifications.Error'
+                            ),
+                        ]
+                    ),
+                    new CleanHtml(),
+                    new TypedRegex([
+                        'type' => 'message',
+                    ]),
+                    new Length(
+                        [
+                            'max' => AddressConstraint::MAX_OTHER_LENGTH,
+                            'maxMessage' => $this->translator->trans(
+                                'This field cannot be longer than %limit% characters',
+                                ['%limit%' => AddressConstraint::MAX_OTHER_LENGTH],
+                                'Admin.Notifications.Error'
+                            ),
+                        ]
+                    ),
+                ],
+            ]);
+
+        $builder->addEventSubscriber(new CustomerAddressFormSubscriber($this->stateChoiceProvider));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function configureOptions(OptionsResolver $resolver)
+    {
+        $resolver->setDefaults(
+            [
+                'validation_groups' => function () {
+                    $groups = ['Default'];
+
+                    //TODO add required fields as validation groups
```

Comment:
@matks i have a question about adding new dataProvider for this case. In controller i use queryBus to fetch address required fields. It is a simple array of strings with names of required fields. For this form i need this array to add validation groups and required flags based on the same array. Would it be a good idea to create a service `RequiredFieldsProvider` with interface witch runs the same query and this service here and at the Controller level to get required fields? Or get rid of validation groups and pass required values as form options and use if statements to determine if fields are required? What do you think?

Reply:
So, we should try to avoid dispatching too many queries **for the same HTTP request**.

CQRS ideal usecase is: "you receive one HTTP request, you dispatch 1 query and/or 1 command". That's the ideal usecase. One usecase = 1 query and/or 1 command. So if it's 1 query, then we're supposed to ask, in the query, for *all the data we need to display* and get it all together.

That is why I suggest the following, which is based on your idea:
- create this RequiredFieldsProvider
- use this RequiredFieldsProvider inside GetCustomerAddressForEditingHandler and return the required fields data into the EditableCustomerAddress so that AddressFormDataProvider can retrieve it and pass this data to the form so it can be validated correctly
- you can also use this RequiredFieldsProvider for the 2nd query (the ones you use in controller to get the data)

Is this alright for you ? If we perform a GetCustomerAddressForEditing query, it means we're asking the Domain for "all the data I need to perform an editing operation for customer address". If the "required fields" are needed to perform an editing operation they should be part of the Query response.

Reply:
Required fields are also needed for create operation so returning them with EditableCustomerAddress only solves half the problem. Or maybe for edit form i could get all required fields with EditableCustomerAddress and for create i can pass required fields as data to form?